src/emerge.cpp

```diff
@@ -183,14 +194,67 @@ EmergeManager::~EmergeManager()
 }
 
 
+BiomeManager *EmergeManager::getWritableBiomeManager()
+{
+	FATAL_ERROR_IF(!m_mapgens.empty(),
+		"Writable managers can only be returned before mapgen init");
+	return biomemgr;
+}
+
+OreManager *EmergeManager::getWritableOreManager()
+{
+	FATAL_ERROR_IF(!m_mapgens.empty(),
+		"Writable managers can only be returned before mapgen init");
+	return oremgr;
+}
+
+DecorationManager *EmergeManager::getWritableDecorationManager()
+{
+	FATAL_ERROR_IF(!m_mapgens.empty(),
+		"Writable managers can only be returned before mapgen init");
+	return decomgr;
+}
+
+SchematicManager *EmergeManager::getWritableSchematicManager()
+{
+	FATAL_ERROR_IF(!m_mapgens.empty(),
+		"Writable managers can only be returned before mapgen init");
+	return schemmgr;
+}
+
+
 void EmergeManager::initMapgens(MapgenParams *params)
 {
 	FATAL_ERROR_IF(!m_mapgens.empty(), "Mapgen already initialised.");
 
 	mgparams = params;
 
-	for (u32 i = 0; i != m_threads.size(); i++)
-		m_mapgens.push_back(Mapgen::createMapgen(params->mgtype, params, this));
+	for (u32 i = 0; i != m_threads.size(); i++) {
+		EmergeParams *p = createEmergeParams();
+		infostream << "EmergeManager: Created params " << p
+			<< " for thread " << i << std::endl;
+		m_mapgens.push_back(Mapgen::createMapgen(params->mgtype, params, p));
+	}
+}
+
+
+EmergeParams *EmergeManager::createEmergeParams()
+{
+	EmergeParams *p = new EmergeParams();
+	assert(p);
+
+	p->ndef = ndef;
+	p->enable_mapgen_debug_info = enable_mapgen_debug_info;
+
+	p->gen_notify_on = gen_notify_on;
+	p->gen_notify_on_deco_ids = &gen_notify_on_deco_ids;
+
+	p->biomemgr = biomemgr->clone();
```

Comment:
i suggest that the pointers on these objects become just objects on stack and implement a copy operator instead of a cloning pointer interface.